sklearn/model_selection/_search_successive_halving.py

```diff
@@ -50,7 +50,11 @@ def _top_k(results, k, itr):
         for a in (results["iter"], results["mean_test_score"], results["params"])
     )
     iter_indices = np.flatnonzero(iteration == itr)
-    sorted_indices = np.argsort(mean_test_score[iter_indices])
+    scores = mean_test_score[iter_indices]
+    # argsort() places NaNs at the end of the array so we move NaNs to the
+    # front of the array so the last `k` items are the those with the
+    # highest scores.
+    sorted_indices = np.roll(np.argsort(scores), np.count_nonzero(np.isnan(scores)))
```

Comment:
I would go as far as filtering out all the nans. For me, it's strange to select candidates that failed in the current iteration for the next iteration.

WDYT @NicolasHug ?

Reply:
I guess it *may* happen that a candidate returns a nan at iteration `n` (because it has too few resources / samples) and returns a non-nan at iteration `n + 1`. So maybe the current code makes sense?

On top of that,  in most cases nan candidates will be filtered out naturally because they won't be part of the "top accuracies" anyway. E.g. when `factor=3` as in the default, it would take at least 66.7% of the candidates to be nans for at least one nan to survive and make it to the next round.

Reply:
I'm not sure what the right thing to do is.

My natural instinct would be to remove candidates with NaNs because "well, you broke this time, so how can you be any good? You are out."

However, it could be that a particular candidate got unlucky in this iteration and we should keep it (Nicolas' `n+1` argument). The other question we'd need to solve if we remove candidates with NaNs is "what do we do if there aren't enough not-NaN candidates left?"

So maybe we should keep them and rely on the fact that for well defined problems(?) there will be at least `k` candidates that aren't NaNs so it doesn't make a difference if we remove them or not.

-> right now I'm leaning towards leaving them in

Reply:
Okay, lets leave them in.